File: packages/grove/src/invariants.jl
const WIP_LIMIT_DEFAULT = 2

function i1_dor_on_progress(st::State)::Vector{String}
    out = String[]
    for w in listnodes(st, :w)
        w.status === :progress || continue
        dor(st, w; pin_coverage=true) || push!(out, "I1: $(w.id) is `progress` but DoR ≢ ⊤")
    end
    out
end

function i2_spike_outputs(st::State)::Vector{String}
    out = String[]
    for w in listnodes(st, :w)
        (w.type === :spike && w.status === :done) || continue
        has_any = any(e -> e.label === :produces && e.from == w.id, st.edges)
        has_any || push!(out,
            "I2: $(w.id) is a done spike but `produces` is empty (no outgoing `produces` edges)")
    end
    out
end

function i3_done_has_evidence(st::State)::Vector{String}
    out = String[]
    for w in listnodes(st, :w)
        w.status === :done || continue
        ev = get(w.fields, :evidence, String[])
        isempty(ev) && push!(out, "I3: $(w.id) is `done` but `evidence` is empty")
    end
    out
end

function i4_wip_limit(st::State; limit::Int=WIP_LIMIT_DEFAULT)::Vector{String}
    n = count(w -> w.status === :progress, listnodes(st, :w))
    n > limit ? ["I4: WIP $(n) exceeds limit $(limit)"] : String[]
end

function i5_blocks_terminal(st::State)::Vector{String}
    out = String[]
    for w in listnodes(st, :w)
        w.status === :progress || continue
        for p in blocked_by(st, w.id)
            np = get(st.nodes, p, nothing)
            if np === nothing
                push!(out, "I5: $(w.id) blocked by missing $p"); continue
            end
            clears_blocks_predecessor(np) ||
                push!(out, "I5: $(w.id) is `progress` but blocker $(p) ($(np.status)) does not satisfy blocks clearance (goals must be verified)")
        end
    end
    out
end

function i7_blocks_dag(st::State)::Vector{String}
    succ = Dict{String,Vector{String}}()
    nodes = Set{String}()
    for e in st.edges
        e.label === :blocks || continue
        push!(get!(succ, e.from, String[]), e.to)
        push!(nodes, e.from); push!(nodes, e.to)
    end
    indeg = Dict{String,Int}(id => 0 for id in nodes)
    for (_, vs) in succ, v in vs
        indeg[v] = get(indeg, v, 0) + 1
    end
    q = [id for (id, d) in indeg if d == 0]
    visited = 0
    while !isempty(q)
        x = pop!(q)
        visited += 1
        for s in get(succ, x, String[])
            indeg[s] -= 1
            indeg[s] == 0 && push!(q, s)
        end
    end
    visited == length(nodes) ? String[] : ["I7: blocks graph contains a cycle"]
end

function i9_feature_bchain(st::State)::Vector{String}
    out = String[]
    for w in listnodes(st, :w)
        (w.type === :feature && w.status in (:ready, :progress)) || continue
        for b in bchain(st, w)
            n = get(st.nodes, b, nothing)
            n === nothing && continue
            n.status in (:validated, :invalidated_acceptable) ||
                push!(out, "I9: $(w.id) is `$(w.status)` but $(b) is `$(n.status)`")
        end
    end
    out
end

function i10_done_fitness(st::State)::Vector{String}
    out = String[]
    for w in listnodes(st, :w)
        w.status === :done || continue
        gs = get(w.fields, :goals, String[])
        f = get(w.fields, :fitness, Dict{String,Int}())
        for g in gs
            haskey(f, g) || push!(out, "I10: $(w.id) is `done` but no fitness delta for $g")
        end
    end
    out
end

function i11_progress_has_session_claim(st::State)::Vector{String}
    out = String[]
    for w in listnodes(st, :w)
        w.status === :progress || continue
        progress_has_session_record(w) ||
            push!(out, "I11: $(w.id) is `progress` but has no session token")
    end
    out
end

function check_orphan_edges(st::State)::Vector{String}
    out = String[]
    for e in st.edges
        haskey(st.nodes, e.from) || push!(out, "edge endpoint missing: $(e.from)")
        haskey(st.nodes, e.to) || push!(out, "edge endpoint missing: $(e.to)")
    end
    out
end

function check_edge_types(st::State)::Vector{String}
    out = String[]
    for e in st.edges
        from = get(st.nodes, e.from, nothing)
        to = get(st.nodes, e.to, nothing)
        (from === nothing || to === nothing) && continue
        ok = if e.label === :blocks
            to.kind === :w
        elseif e.label === :causes
            from.kind === :t && to.kind === :w
        elseif e.label === :implements
            from.kind === :w && to.kind === :d
        elseif e.label === :asks
            from.kind === :q
        elseif e.label === :tests
            from.kind === :b && to.kind === :q
        elseif e.label === :targets
            from.kind === :b && to.kind === :w
        elseif e.label === :produces
            from.kind === :w && to.kind in (:d, :q, :b, :y)
        elseif e.label === :supersedes
            (from.kind === :d && to.kind === :d) || (from.kind === :y && to.kind === :y)
        elseif e.label === :distills
            from.kind === :y && to.kind in (:d, :q, :b)
        else
            false
        end
        ok || push!(out, "edge type mismatch: $(e.from) -$(e.label)-> $(e.to)")
    end
    out
end

function discovery_anchor_issues(st::State, x::Node)::Vector{String}
    out = String[]
    x.kind === :y || return out
    x.archived && return out
    has_origin = false
    for e in st.edges
        if e.label === :produces && e.to == x.id
            src = get(st.nodes, e.from, nothing)
            src !== nothing && src.kind === :w && (has_origin = true)
        elseif e.label === :distills && e.from == x.id
            dst = get(st.nodes, e.to, nothing)
            dst !== nothing && dst.kind in (:d, :q, :b) && (has_origin = true)
        end
        has_origin && break
    end
    has_origin || push!(out,
        "I12: $(x.id) has no provenance edge (needs `produces` from a W or `distills` to a D/Q/B)")
    surface = get(x.fields, :surface, String[])
    why = get(x.fields, :why, String[])
    (isempty(surface) && !prose_field_nonempty(why)) && push!(out,
        "I12: $(x.id) has empty `surface` and empty `why` (≥1 anchor required)")
    tags = get(x.fields, :tags, String[])
    isempty(tags) && push!(out, "I12: $(x.id) has empty `tags` (≥1 glossary term required)")
    out
end

function check_discovery_anchors(st::State)::Vector{String}
    out = String[]
    for x in listnodes(st, :y)
        append!(out, discovery_anchor_issues(st, x))
    end
    out
end

function check_area_membership(st::State)::Vector{String}
    out = String[]
    for g in listnodes(st, :g; include_archived=true)
        aref = get(g.fields, :area, "")
        if !(aref isa AbstractString) || isempty(strip(String(aref)))
            push!(out, "I13: $(g.id) has no `area` field (every goal belongs to an area: `grove set $(g.id) area=A-NN`)")
            continue
        end
        zn = get(st.nodes, String(aref), nothing)
        (zn === nothing || zn.kind !== :a) &&
            push!(out, "I13: $(g.id) area $aref does not reference an existing area (a) node")
    end
    out
end

function check_all(st::State)::Vector{String}
    vcat(
        i1_dor_on_progress(st),
        i2_spike_outputs(st),
        i3_done_has_evidence(st),
        i4_wip_limit(st),
        i5_blocks_terminal(st),
        i7_blocks_dag(st),
        i9_feature_bchain(st),
        i10_done_fitness(st),
        i11_progress_has_session_claim(st),
        check_orphan_edges(st),
        check_edge_types(st),
        check_discovery_anchors(st),
        check_area_membership(st),
    )
end

"""Return `nothing` on success, otherwise an error string (edge not added)."""
function validate_and_push_edge!(
    st::State, from::AbstractString, label::Symbol, to::AbstractString;
    bump_nodes::Bool=true,
)::Union{Nothing,String}
    from = String(strip(from))
    to = String(strip(to))
    label in EDGE_LABELS || return "unknown edge label: $(label)"
    haskey(st.nodes, from) || return "missing node $(from)"
    haskey(st.nodes, to) || return "missing node $(to)"
    if any(e -> e.from == from && e.label === label && e.to == to, st.edges)
        return nothing
    end
    e = Edge(from, label, to)
    push!(st.edges, e)
    stamp_new_edge!(e)
    if label === :blocks && !isempty(i7_blocks_dag(st))
        pop!(st.edges)
        return "I7: blocks introduces a cycle"
    end
    et = check_edge_types(st)
    if !isempty(et)
        pop!(st.edges)
        return et[end]
    end
    if bump_nodes
        stamp_touch_node!(getnode(st, from))
        stamp_touch_node!(getnode(st, to))
    end
    nothing
end

File: packages/core/src/invariants.rs
use crate::algebra::{bchain, blocked_by};
use crate::dor::dor;
use crate::model::{Edge, Kind, Node, State};
use crate::session::progress_has_session_record;
use crate::status::{clears_blocks_predecessor, listnodes, prose_field_nonempty, EDGE_LABELS};
use crate::times::{stamp_new_edge, stamp_touch_node};
use std::collections::{BTreeMap, BTreeSet};

pub const WIP_LIMIT_DEFAULT: i64 = 2;

pub fn i1_dor_on_progress(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for w in listnodes(st, Kind::W, false) {
        if w.status != "progress" {
            continue;
        }
        if !dor(st, w, true) {
            out.push(format!("I1: {} is `progress` but DoR ≢ ⊤", w.id));
        }
    }
    out
}

pub fn i2_spike_outputs(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for w in listnodes(st, Kind::W, false) {
        if !(w.wtype.as_deref() == Some("spike") && w.status == "done") {
            continue;
        }
        let has_any = st
            .edges
            .iter()
            .any(|e| e.label == "produces" && e.from == w.id);
        if !has_any {
            out.push(format!(
                "I2: {} is a done spike but `produces` is empty (no outgoing `produces` edges)",
                w.id
            ));
        }
    }
    out
}

pub fn i3_done_has_evidence(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for w in listnodes(st, Kind::W, false) {
        if w.status != "done" {
            continue;
        }
        if w.lines("evidence").is_empty() {
            out.push(format!("I3: {} is `done` but `evidence` is empty", w.id));
        }
    }
    out
}

pub fn i4_wip_limit(st: &State) -> Vec<String> {
    i4_wip_limit_with(st, WIP_LIMIT_DEFAULT)
}

pub fn i4_wip_limit_with(st: &State, limit: i64) -> Vec<String> {
    let n = listnodes(st, Kind::W, false)
        .iter()
        .filter(|w| w.status == "progress")
        .count() as i64;
    if n > limit {
        vec![format!("I4: WIP {} exceeds limit {}", n, limit)]
    } else {
        Vec::new()
    }
}

pub fn i5_blocks_terminal(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for w in listnodes(st, Kind::W, false) {
        if w.status != "progress" {
            continue;
        }
        for p in blocked_by(st, &w.id) {
            match st.nodes.get(&p) {
                None => {
                    out.push(format!("I5: {} blocked by missing {}", w.id, p));
                    continue;
                }
                Some(np) => {
                    if !clears_blocks_predecessor(np) {
                        out.push(format!(
                            "I5: {} is `progress` but blocker {} ({}) does not satisfy blocks clearance (goals must be verified)",
                            w.id, p, np.status
                        ));
                    }
                }
            }
        }
    }
    out
}

pub fn i7_blocks_dag(st: &State) -> Vec<String> {
    let mut succ: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let mut nodes: BTreeSet<String> = BTreeSet::new();
    for e in &st.edges {
        if e.label != "blocks" {
            continue;
        }
        succ.entry(e.from.clone()).or_default().push(e.to.clone());
        nodes.insert(e.from.clone());
        nodes.insert(e.to.clone());
    }
    let mut indeg: BTreeMap<String, i64> = nodes.iter().map(|id| (id.clone(), 0)).collect();
    for vs in succ.values() {
        for v in vs {
            *indeg.entry(v.clone()).or_insert(0) += 1;
        }
    }
    let mut q: Vec<String> = indeg
        .iter()
        .filter(|(_, d)| **d == 0)
        .map(|(id, _)| id.clone())
        .collect();
    let mut visited = 0usize;
    while let Some(x) = q.pop() {
        visited += 1;
        if let Some(ss) = succ.get(&x) {
            for s in ss {
                if let Some(d) = indeg.get_mut(s) {
                    *d -= 1;
                    if *d == 0 {
                        q.push(s.clone());
                    }
                }
            }
        }
    }
    if visited == nodes.len() {
        Vec::new()
    } else {
        vec!["I7: blocks graph contains a cycle".to_string()]
    }
}

pub fn i9_feature_bchain(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for w in listnodes(st, Kind::W, false) {
        if !(w.wtype.as_deref() == Some("feature")
            && matches!(w.status.as_str(), "ready" | "progress"))
        {
            continue;
        }
        for b in bchain(st, w) {
            let n = match st.nodes.get(&b) {
                Some(n) => n,
                None => continue,
            };
            if !matches!(n.status.as_str(), "validated" | "invalidated_acceptable") {
                out.push(format!(
                    "I9: {} is `{}` but {} is `{}`",
                    w.id, w.status, b, n.status
                ));
            }
        }
    }
    out
}

pub fn i10_done_fitness(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for w in listnodes(st, Kind::W, false) {
        if w.status != "done" {
            continue;
        }
        let f = w.fitness();
        for g in w.lines("goals") {
            if !f.contains_key(&g) {
                out.push(format!(
                    "I10: {} is `done` but no fitness delta for {}",
                    w.id, g
                ));
            }
        }
    }
    out
}

pub fn i11_progress_has_session_claim(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for w in listnodes(st, Kind::W, false) {
        if w.status != "progress" {
            continue;
        }
        if !progress_has_session_record(w) {
            out.push(format!(
                "I11: {} is `progress` but has no session token",
                w.id
            ));
        }
    }
    out
}

pub fn check_orphan_edges(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for e in &st.edges {
        if !st.nodes.contains_key(&e.from) {
            out.push(format!("edge endpoint missing: {}", e.from));
        }
        if !st.nodes.contains_key(&e.to) {
            out.push(format!("edge endpoint missing: {}", e.to));
        }
    }
    out
}

pub fn check_edge_types(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for e in &st.edges {
        let from = match st.nodes.get(&e.from) {
            Some(n) => n,
            None => continue,
        };
        let to = match st.nodes.get(&e.to) {
            Some(n) => n,
            None => continue,
        };
        let ok = match e.label.as_str() {
            "blocks" => to.kind == Kind::W,
            "causes" => from.kind == Kind::T && to.kind == Kind::W,
            "implements" => from.kind == Kind::W && to.kind == Kind::D,
            "asks" => from.kind == Kind::Q,
            "tests" => from.kind == Kind::B && to.kind == Kind::Q,
            "targets" => from.kind == Kind::B && to.kind == Kind::W,
            "produces" => {
                from.kind == Kind::W && matches!(to.kind, Kind::D | Kind::Q | Kind::B | Kind::Y)
            }
            "supersedes" => {
                (from.kind == Kind::D && to.kind == Kind::D)
                    || (from.kind == Kind::Y && to.kind == Kind::Y)
            }
            "distills" => from.kind == Kind::Y && matches!(to.kind, Kind::D | Kind::Q | Kind::B),
            _ => false,
        };
        if !ok {
            out.push(format!(
                "edge type mismatch: {} -{}-> {}",
                e.from, e.label, e.to
            ));
        }
    }
    out
}

pub fn discovery_anchor_issues(st: &State, x: &Node) -> Vec<String> {
    let mut out = Vec::new();
    if x.kind != Kind::Y {
        return out;
    }
    if x.archived {
        return out;
    }
    let mut has_origin = false;
    for e in &st.edges {
        if e.label == "produces" && e.to == x.id {
            if let Some(src) = st.nodes.get(&e.from) {
                if src.kind == Kind::W {
                    has_origin = true;
                }
            }
        } else if e.label == "distills" && e.from == x.id {
            if let Some(dst) = st.nodes.get(&e.to) {
                if matches!(dst.kind, Kind::D | Kind::Q | Kind::B) {
                    has_origin = true;
                }
            }
        }
        if has_origin {
            break;
        }
    }
    if !has_origin {
        out.push(format!(
            "I12: {} has no provenance edge (needs `produces` from a W or `distills` to a D/Q/B)",
            x.id
        ));
    }
    if x.lines("surface").is_empty() && !prose_field_nonempty(&x.lines("why")) {
        out.push(format!(
            "I12: {} has empty `surface` and empty `why` (≥1 anchor required)",
            x.id
        ));
    }
    if x.lines("tags").is_empty() {
        out.push(format!(
            "I12: {} has empty `tags` (≥1 glossary term required)",
            x.id
        ));
    }
    out
}

pub fn check_discovery_anchors(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for x in listnodes(st, Kind::Y, false) {
        out.extend(discovery_anchor_issues(st, x));
    }
    out
}

pub fn check_area_membership(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    for g in listnodes(st, Kind::G, true) {
        let aref = g.single("area");
        if aref.trim().is_empty() {
            out.push(format!(
                "I13: {} has no `area` field (every goal belongs to an area: `grove set {} area=A-NN`)",
                g.id, g.id
            ));
            continue;
        }
        match st.nodes.get(&aref) {
            Some(zn) if zn.kind == Kind::A => {}
            _ => out.push(format!(
                "I13: {} area {} does not reference an existing area (a) node",
                g.id, aref
            )),
        }
    }
    out
}

pub fn check_all(st: &State) -> Vec<String> {
    let mut out = Vec::new();
    out.extend(i1_dor_on_progress(st));
    out.extend(i2_spike_outputs(st));
    out.extend(i3_done_has_evidence(st));
    out.extend(i4_wip_limit(st));
    out.extend(i5_blocks_terminal(st));
    out.extend(i7_blocks_dag(st));
    out.extend(i9_feature_bchain(st));
    out.extend(i10_done_fitness(st));
    out.extend(i11_progress_has_session_claim(st));
    out.extend(check_orphan_edges(st));
    out.extend(check_edge_types(st));
    out.extend(check_discovery_anchors(st));
    out.extend(check_area_membership(st));
    out
}

pub fn validate_and_push_edge(
    st: &mut State,
    from: &str,
    label: &str,
    to: &str,
    bump_nodes: bool,
) -> Option<String> {
    let from = from.trim().to_string();
    let to = to.trim().to_string();
    if !EDGE_LABELS.contains(&label) {
        return Some(format!("unknown edge label: {}", label));
    }
    if !st.nodes.contains_key(&from) {
        return Some(format!("missing node {}", from));
    }
    if !st.nodes.contains_key(&to) {
        return Some(format!("missing node {}", to));
    }
    if st
        .edges
        .iter()
        .any(|e| e.from == from && e.label == label && e.to == to)
    {
        return None;
    }
    let mut e = Edge::new(&from, label, &to);
    stamp_new_edge(&mut e);
    st.edges.push(e);
    if label == "blocks" && !i7_blocks_dag(st).is_empty() {
        st.edges.pop();
        return Some("I7: blocks introduces a cycle".to_string());
    }
    let et = check_edge_types(st);
    if !et.is_empty() {
        st.edges.pop();
        return et.into_iter().last();
    }
    if bump_nodes {
        if let Some(n) = st.nodes.get_mut(&from) {
            stamp_touch_node(n);
        }
        if let Some(n) = st.nodes.get_mut(&to) {
            stamp_touch_node(n);
        }
    }
    None
}
